### PR TITLE
Add simpleclient_executor

### PR DESCRIPTION
**Pull request by cenkakin**
Support for instrumented executor and scheduled executor services

### CODE REVIEW

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+  private final ExecutorService delegate;
+  private final Counter.Child submittedCounter;
+  private final Counter.Child completedCounter;
+  private final Gauge.Child runningGauge;
+  private final Histogram.Child durationHistogram;
+
+  static {
+    ST_COUNTER = Counter.build()
+        .name("instrumented_executor_service_submitted_count")
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: Counters end with _total by convention, _count is used by Summaries/Histograms

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        .help("Instrumented executor service submitted total count")
+        .register();
+    CT_COUNTER = Counter.build()
+        .name("instrumented_executor_service_completed_count")
+        .labelNames("serviceName")
+        .help("Instrumented executor service completed total count")
+        .register();
+
+    R_GAUGE = Gauge.build()
+        .name("instrumented_executor_service_running_count")
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: Gauges shouldn't have a suffix

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  private final Histogram.Child durationHistogram;
+
+  static {
+    ST_COUNTER = Counter.build()
+        .name("instrumented_executor_service_submitted_count")
+        .labelNames("serviceName")
+        .help("Instrumented executor service submitted total count")
+        .register();
+    CT_COUNTER = Counter.build()
+        .name("instrumented_executor_service_completed_count")
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: The instrumented and executor seem kinda redundant here

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        .help("Instrumented executor service completed total count")
+        .register();
+
+    R_GAUGE = Gauge.build()
+        .name("instrumented_executor_service_running_count")
+        .labelNames("serviceName")
+        .help("Instrumented executor service running total count")
+        .register();
+
+    D_HISTOGRAM = Histogram.build()
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: A histogram has buckets to cover 1ms-10s by default, is that appropriate for this sort of usage?

A quantile-less summary is probably a better choice.

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class InstrumentedExecutorService implements ExecutorService {
+
+  private static final AtomicLong NAME_COUNTER = new AtomicLong();
+  private static final Counter ST_COUNTER;
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: You can do this all in one line, ST_ isn't very descriptive.

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  private final ExecutorService delegate;
+  private final Counter.Child submittedCounter;
+  private final Counter.Child completedCounter;
+  private final Gauge.Child runningGauge;
+  private final Histogram.Child durationHistogram;
+
+  static {
+    ST_COUNTER = Counter.build()
+        .name("instrumented_executor_service_submitted_count")
+        .labelNames("serviceName")
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: camel_case is the convention for label names, `executor` would be a better name

File: simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    D_HISTOGRAM = Histogram.build()
+        .name("instrumented_executor_service_duration_seconds")
+        .labelNames("serviceName")
+        .help("Instrumented executor service duration (s)")
+        .register();
+  }
+
+
+  public InstrumentedExecutorService(ExecutorService delegate) {
+    this(delegate, "instrumented_delegate_" + NAME_COUNTER.incrementAndGet());
```
Comment (brian-brazil) [inline simpleclient_executor/src/main/java/io/prometheus/client/executor/InstrumentedExecutorService.java]: unknown_...

would be better. No point in making it longer than it has to be.

**Comment by cenkakin** [comment]
Can you review PR again?

**Comment by brian-brazil** [comment]
My high level comments still apply, I don't think this belongs here.

**Comment by cemo** [comment]
I can not follow your comment. I can only see an instrumented implementation of executor services. CodaHale's metric has also implementation of instrumented executor services. I think that providing parity between these two project would be nice. What do you think? 

Disclosure: I am Cenk's colleague.

**Comment by brian-brazil** [comment]
This is only one executor, what if someone already has an executor? This doesn't help them.

I don't think it's sane or useful to implement every possible Java interface.

**Comment by cemo** [comment]
This is an not ordinary java interface. `ExecutorService` is one of the fundamental class in java concurrency package.  It is used almost all mainstream libraries such as `Netty`, `Spring`, `Tomcat`, `Guava`, `Jetty`, `Logback`. Their usage is not limited with simple things. They have extensive support for implementing concurrency related things around them. I think that `ExecutorService` is good asset for the project.

**Comment by brian-brazil** [comment]
You didn't address my point.

**Comment by cemo** [comment]
We are addressing the most common usage of `Executor` as it was done in other libraries. Even in the `java.util.concurrent.Executors` package, methods are returning `ExecutorService`. Guava has helper class `ForwardingExecutorService` for `ExecutorService` but not `Executor`. 

If someone needs to use an `Executor`, it will be mostly based on `AbstractExecutorService` or another useful class which is implementing `ExecutorService`.

So for the people who are using `Executor` which is not a subclass of `ExecutorService`, we have not so much gems for them. If they want to contribute, here is the code. They can improve it. We are trying to cover as much as people, as `Guava` did.

**Comment by checketts** [comment]
I can see the value of having this in the client as I know it has been very painful for me to monitor my current executors. That being said, I didn't have much success with the Drop Wizard equivalent back when I was using it since it would have forced me to use that specific implementation.

How far could we get with an a collector can take and existing executor and expose metrics for it?

Alternatively (if you are using Spring) would using `TaskDecorator` be a good extension point? (see ThreadPoolTaskExecutor)

**Comment by cemo** [comment]
@brian-brazil we are considering to release a new library named `client_java_extra` and will add new modules there. What do you think?

**Comment by brian-brazil** [comment]
That naming might be a bit confusing, but you're free to release what you like.

**Pull request closed without merging**